lib/cache/cache.go

```diff
@@ -207,8 +239,92 @@ type Cache struct {
 	closedFlag int32
 }
 
+func (c *Cache) setInitError(err error) {
+	c.initOnce.Do(func() {
+		c.initErr = err
+		close(c.initC)
+	})
+}
+
+// setReadOK updates Cache.ok, which determines whether the
+// cache is accessible for reads.
+func (c *Cache) setReadOK(ok bool) {
+	if ok == c.getReadOK() {
```

Comment:
Both; we could theoretically have hundreds or even thousands of concurrent reads.  Because of this, the write lock is grabbed worst-case twice per full reset of cache state.